### ISSUE TITLE
Audio devices can have duplicate names and will cause options to crash

### ISSUE THREAD

**Issue by Francesco149** [ppy/osu#605]
On my linux gentoo x64 system, the latest revision of ozu!lazer crashes upon clicking the settings menu.

This seems to be caused by the game attempting to add duplicate device names here: https://github.com/ppy/osu/blob/master/osu.Game/Overlays/Options/Sections/Audio/AudioDevicesOptions.cs#L36

I tried printing out the device list it tries to add, and it's:

```Default
Loopback: Loopback PCM
Loopback: Loopback PCM
HDA Intel PCH: ALC1150 Analog
HDA Intel PCH: ALC1150 Digital
HDA ATI HDMI: Generic Digital
PulseAudio Sound Server
```

Now, I have no idea how my system has two sound loopback devices with the same name, but because it is possible, I would recommend either ignoring the duplicate device name or changing it to something like "name (2)" to prevent crashes on such a setup. You could also not use the device names as the keys for the deviceItems map.

This is an example that would ignore duplicate device names:
```
            foreach (var d in audio.AudioDeviceNames)                                                               
            {                                                                                                       
                if (deviceItems.All(kv => kv.Key != d))                                                             
                    deviceItems.Add(new KeyValuePair<string, string>(d, d));                                        
            }     
```

Full stack-trace of the crash: https://gist.github.com/Francesco149/b15808d5661ddd82025864d70ef85b23

Here is additional info on my loopback devices from aplay:

```$ aplay -L
null
    Discard all samples (playback) or generate zero samples (capture)
pulse
    PulseAudio Sound Server
default
    PulseAudio Sound Server
sysdefault:CARD=Loopback
    Loopback, Loopback PCM
    Default Audio Device
front:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    Front speakers
surround21:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    4.0 Surround output to Front and Rear speakers
surround41:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    5.0 Surround output to Front, Center and Rear speakers
surround51:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=Loopback,DEV=0
    Loopback, Loopback PCM
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
```

**Comment by ZLima12** [comment]
You seem to have quite a few ideas, why not try them out and submit a PR?

**Comment by Francesco149** [comment]
Well, I'd like to agree on the preferred method to handle this first. I tested out the simple example that just ignores duplicates and it works.

**Comment by peppy** [comment]
Thanks for the detailed report. We'll need to figure out whether we want to ignore duplicates or include them all. I'd be especially interested in whether this can also happen on windows.

**Comment by Francesco149** [comment]
I googled around a bit and it seems that at least for recording device it's possible to have non-unique names on some versions on windows and they can cause issues: https://support.microsoft.com/en-us/help/928236/recording-devices-may-not-work-if-the-devices-have-the-same-name-in-windows-vista

**Comment by Tom94** [comment]
Yes, non-unique names are possible. In the current implementation we decided to not handle them, because it involved sophisticated OS-specific code to obtain unique identifiers. BASS only provided a way to select devices based on name IIRC.

**Comment by peppy** [comment]
If it becomes an issue for end-users, we can likely request support of bass at that point. For now we should probably address the crash and leave a comment about non-unique identifiers.

**Comment by Tom94** [comment]
@peppy To answer your question: I managed to reproduce on windows. Simply rename 2 audio devices that come from the same physical device such that their names coincide.
![image of duplicate name](http://i.imgur.com/cpa5WZ7.png)
In this case I simply renamed the digital out of my on-board soundcard.

**Comment by peppy** [comment]
Resolved by #606 for the time being. Please comment/reopen if this issue affects your ability to choose the device you want (and you have no way of fixing this).
